src/panoptes/pocs/camera/camera.py

```diff
@@ -636,6 +664,97 @@ def get_thumbnail(self, seconds, file_path, thumbnail_size, keep_file=False, *ar
+
+        # Define an inner function to run in a thread
+        def check_temp(required_stable_time, sleep_delay, timeout):
+            if required_stable_time > timeout:
+                raise ValueError("required_stable_time must be less than timeout.")
+            if sleep_delay > timeout:
+                raise ValueError("sleep_delay must be less than timeout.")
+
+            # Wait until stable temperature persists or timeout
+            time_stable = 0
+            timer = CountdownTimer(duration=timeout)
+            while True:
+                if timer.expired():
+                    break
+                # We may need to restart the thread before it has finished
+                if self._restart_temperature_thread:
+                    self._restart_temperature_thread = False
+                    return
+                # Check if the temperature is within tolerance
+                if abs(self.temperature - self.target_temperature) < self.temperature_tolerance:
+                    time_stable += sleep_delay
+                    if time_stable >= required_stable_time:
+                        self.logger.info(f"Temperature has stabilised on {self}.")
+                        self._is_temperature_stable = True
+                        return
+                else:
+                    time_stable = 0  # Reset the countdown
+                # timer.sleep(max_delay=sleep_delay)  # This hangs for some reason
+                time.sleep(sleep_delay)
+            raise error.Timeout(f"Timeout while waiting for stable temperture on {self}.")
+
+        # Restart countdown if one is already in progress
+        if self._temperature_thread is not None:
+            if self._temperature_thread.is_alive():
+                self.logger.warning(f"Attempted to wait for stable temperature on {self}"
+                                    " while wait already in progress. Restarting countdown.")
+                self._restart_temperature_thread = True
+                self._temperature_thread.join()
+
+        # Wait for stable temperature
+        self._is_temperature_stable = False
+        self.logger.info(f"Waiting for stable temperature on {self}.")
+        self._temperature_thread = threading.Thread(
+                target=check_temp, args=(required_stable_time, sleep_delay, timeout))
+        self._temperature_thread.start()
+        if blocking:
+            self._temperature_thread.join()
+
+    @abstractmethod
+    def _set_target_temperature(self, target):
+        """Camera-specific function to set the target temperature.
+
+        Note:
+            Each sub-class is required to implement this abstract method. The derived
+            method should at a minimum implement the described parameters.
+
+        Args:
+            target (astropy.units.Quantity): The target temperature.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _set_cooling_enabled(self, enable):
```

Comment:
You don't need to change anything here, but I keep wondering about the logic of having abstract private methods... :thinking: